File: practice/src/main/kotlin/ar/com/miura/jvm/permutations/basic/Permutations.java
package ar.com.miura.jvm.permutations.basic;

public class Permutations {
    public java.util.List<java.util.List<Integer>> permute(int[] nums) {
        java.util.List<java.util.List<Integer>> permutations = new java.util.ArrayList<java.util.List<Integer>>();
        permutations(nums, 0, permutations);
        return permutations;
    }

    private void permutations(int[] nums, int i, java.util.List<java.util.List<Integer>> permutations) {

        int size = nums.length;
        if (i==size-1) {
            permutations.add(convert(nums));
        } else {
            for(int j=i;j<size;j++) {
                swap(i,j, nums);
                permutations(nums, i+1, permutations);
                swap(j,i, nums);
            }
        }
    }

    private void swap(int i,int j, int[] nums) {
        int tempI = nums[i];
        int tempJ = nums[j];
        nums[i] = tempJ;
        nums[j] = tempI;
    }

    private java.util.List<Integer> convert(int[] numbers) {
        return java.util.Arrays.asList(
                java.util.Arrays.stream(numbers).boxed().toArray(Integer[]::new)
        );
    }
}

File: practice/src/main/kotlin/ar/com/miura/lc/backtracking/Permutations.kt
package ar.com.miura.lc.backtracking

class Permutations {

    fun getPermutations(array: List<Int>): List<List<Int>> {
        var permutations = mutableListOf<List<Int>>();
        permutationsHelper(0, array.toMutableList(), permutations);
        return permutations;
    }

    private fun permutationsHelper(i:Int, array:MutableList<Int>, permutations: MutableList<List<Int>>) {
        if (i==array.size-1) {
            permutations.add(array.toMutableList())

        } else {
            var size = array.size;
            for(j in i..(size-1)) {
                swap(array, i, j);
                permutationsHelper(i+1, array, permutations);
                swap(array,j,i);
            }
        }
    }

    private fun swap(array:MutableList<Int>, i:Int, j:Int) {
        var tempI = array.get(i);
        var tempJ = array.get(j);
        array.set(i, tempJ);
        array.set(j, tempI)
    }

}